### ISSUE TITLE
Onnx softplus accuracy problem

### ISSUE THREAD

**Issue by Shawn314** [onnx/onnx#3997]
**Describe the bug**
I found a accuracy problem about onnx softplus. Onnx softplus operator behavior is inconsistent with other framework calculation results.

**Urgency**
None.

**System information**
- OS Platform and Distribution : Linux Ubuntu 18.04
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.10.2
- Python version: 3.8.10

**To Reproduce**
- A scalar is **-12.3392**
- pytorch and tensorflow softplus result is 4.37676e-6, but onnx is 4.41073e-6.

**Expected behavior**
All Inference framework has same result.

**Screenshots**
Pytorch
![image](https://user-images.githubusercontent.com/30689370/152963056-44896553-c3f4-4adb-a16c-c91db668345a.png)
Tensorflow
![image](https://user-images.githubusercontent.com/30689370/152963165-755510e4-1504-48ec-98dc-5d8ba01a256c.png)
Onnx
![image](https://user-images.githubusercontent.com/30689370/152963359-4b8644f4-d904-4ced-970d-c4f7750bd56b.png)

**Additional context**
None

**Comment by jcwchen** [comment]
Hi @Shawn314,
It seems related to floating type (float32 vs float64)in NumPy:
```
np.log(np.exp([-12.3392]).astype(np.float64) + 1)  # NumPy default floating type is float64
#  output: array([4.37675845e-06])

np.log(np.exp([-12.3392]).astype(np.float32) + 1)
#  output: array([4.4107337e-06], dtype=float32)
```
From my experiments, both Tensorflow and PyTorch produced `~4.376758e-06` by giving float32 and float64. Since ONNX backend test uses NumPy to create test cases (input/output), the output result might be slightly different due to different behavior among Numpy, Tensorflow, PyTorch, although they use the same formula to compute result.

Another finding:
```
tf.math.log(tf.exp(tf.Variable([-12.3392], dtype=tf.float64)) + 1)
#  output: <tf.Tensor: shape=(1,), dtype=float64, numpy=array([4.37675845e-06])>
tf.math.log(tf.exp(tf.Variable([-12.3392], dtype=tf.float32)) + 1)
#  output: <tf.Tensor: shape=(1,), dtype=float32, numpy=array([4.410734e-06], dtype=float32)>
```
If we use TensorFlow with [Softplus's formula](<tf.Tensor: shape=(1,), dtype=float32, numpy=array([4.410734e-06], dtype=float32)>): `log(exp(x) + 1)`. The result will look more consistent as what NumPy does...

Then I found actually numpy.log does not perform accurately when x is very small for `np.log(x+1)`. Using `np.log1p(x)` would be more accurate: https://stackoverflow.com/questions/49538185/purpose-of-numpy-log1p.
```
np.log(np.exp([-12.3392]).astype(np.float32) + 1)
#  output: array([4.4107337e-06], dtype=float32)
np.log1p(np.exp([-12.3392]).astype(np.float32))
#  output: array([4.3767586e-06], dtype=float32)
```

If the current test_softplus (backend/test/data) output is quite different from PyTorch and TensorFlow, we might be able to use `np.log1p(x)` to create a more precise result. Please let me know if there is a obvious gap. Otherwise, for normal x, using np.log(x+1) should be OK.

**Comment by Shawn314** [comment]
Hi @jcwchen 
Thank you reply, very helpful for me. Basically understand the principle.
But how to use np.log1p(x) or np.float64 in onnx softplus caculation to produce accuracy result for me.

I am newer for onnx. Looking forward your reply.
![image](https://user-images.githubusercontent.com/30689370/153117885-4db57d3b-38f7-48d0-9685-5f5fedc759e1.png)

**Comment by jcwchen** [comment]
So you want to get an inference result from ONNX? If so, please use a Runtime which supports ONNX inference. (For instance, ONNXRuntime: https://github.com/microsoft/onnxruntime) ONNX itself doesn't have a Runtime kernel.

To clarify: Since ONNX doesn't have any Runtime kernel, it uses NumPy to compute result for backend data. For example: https://github.com/onnx/onnx/blob/83fa57c74edfd13ddac9548b8a12f9e3e2ed05bd/onnx/backend/test/case/node/softplus.py#L26 Please note that it is not a real Runtime maintained by ONNX. What ONNX cares is using NumPy correctly and precisely enough to compute the right result for backend test data, which can be verified by all Runtimes which support ONNX.

**Comment by Shawn314** [comment]
understand, Thank you @jcwchen. Please close this issue.
